File: chapter03/python/twoSum.py
"""
对于无序的数组只能穷举, 暴力穷举的时间复杂度为O(N^2), 可借助HashMap, 以空间换时间, 因为HashMap的查询为O(1)
"""


def two_sum(nums, target):
    # 时间复杂度O(N), 空间复杂度O(N)
    d = {}
    for i, e in enumerate(nums):
        d[e] = i

    for i, e in enumerate(nums):
        k = target - e
        if k in d and d[k] != i:
            return i, d[k]
    return -1, -1


if __name__ == "__main__":
    nums = [3, 1, 3, 6]
    target = 6
    print(two_sum(nums, target))

File: chapter03/cpp/two_sum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <unordered_map>

using namespace std;

bool two_sum(const vector<int>& nums, const int target)
{   
    unordered_map<int, int> m;
    for(int i=0; i<nums.size(); i++)
        m[nums[i]] = i;
    
    for(int i=0; i<nums.size(); i++)
    {
        int k = target - nums[i];
        if(m.count(k) && m[k] != i)
            return true;
    }
    return false;
}

int main()
{
    vector<int> nums = {3, 1, 2, 4, 6};
    int target = 6;
    cout << two_sum(nums, target) << endl;
    return 0;
}